src/widgets/test-results/loadPatientTestData/helpers.ts

```diff
@@ -0,0 +1,250 @@
+    category: "laboratory",
+    _sort: "-_date",
+    _summary: "data",
+    _format: "json",
+    _count: "1"
+  });
+
+  return (await request.next().value)?.entry?.[0]?.resource?.id;
+};
+
+/**
+ * Retrieves cached user testresults data
+ * Checks the indicator against the backend while doing so
+ *
+ * @param { string } patientUuid
+ * @param { PatientData } data
+ * @param { string } indicator UUID of the newest observation
+ */
+export const getUserDataFromCache = async (
+  patientUuid: string
+): Promise<PatientData | undefined> => {
+  const [data, , indicator] = cacheStore.getState()[patientUuid] || [];
+
+  if (!!data && (await getLatestObsUuid(patientUuid)) === indicator)
+    return data;
+};
+
+/**
+ * Iterator
+ * @param queries
+ */
+function* fhirObservationRequests(queries: Record<string, string>) {
+  const fhirPathname = "/openmrs/ws/fhir2/R4/Observation";
+  const path =
+    fhirPathname +
+    "?" +
+    Object.entries(queries)
+      .map(([q, v]) => q + "=" + v)
+      .join("&");
+
+  const pathWithPageOffset = offset =>
+    path + "&_getpagesoffset=" + offset * pageSize;
+  let offsetCounter = 0;
+  while (true) {
+    yield fetch(pathWithPageOffset(offsetCounter++)).then(res => res.json());
+  }
+}
+
+const CHUNK_PREFETCH_COUNT = 6;
+/**
+ * Load all patient testresult observations in parallel
+ *
+ * @param { string } patientUuid
+ * @returns { Promise<ObsRecord[]> }
+ */
+export const loadObsEntries = async (
+  patientUuid: string
+): Promise<ObsRecord[]> => {
+  const requests = fhirObservationRequests({
+    patient: patientUuid,
+    category: "laboratory",
+    _sort: "-_date",
+    _summary: "data",
+    _format: "json",
+    _count: "" + pageSize
+  });
+
+  let responses = await Promise.all(
+    retrieveFromIterator(requests, CHUNK_PREFETCH_COUNT)
+  );
+
+  const total = responses[0].total;
+
+  if (total > CHUNK_PREFETCH_COUNT * pageSize) {
+    const missingRequestsCount =
+      Math.ceil(total / pageSize) - CHUNK_PREFETCH_COUNT;
+    responses = [
+      ...responses,
+      ...(await Promise.all(
+        retrieveFromIterator(requests, missingRequestsCount)
+      ))
+    ];
+  }
+
+  console.log({ responses });
+
+  return responses
+    .slice(0, Math.ceil(total / pageSize))
+    .flatMap(res => res.entry.map(e => e.resource));
+};
+
+export const getEntryConceptClassUuid = entry => entry.code.coding[0].code;
+
+const conceptCache: Record<ConceptUuid, Promise<ConceptRecord>> = {};
+/**
+ * fetch all concepts for all given observation entries
+ */
+export const loadPresentConcepts = (
+  entries: ObsRecord[]
+): Promise<ConceptRecord[]> =>
+  Promise.all(
+    [...new Set(entries.map(getEntryConceptClassUuid))].map(
+      conceptUuid =>
+        conceptCache[conceptUuid] ||
+        (conceptCache[conceptUuid] = fetch(
+          "/openmrs/ws/rest/v1/concept/" + conceptUuid + "?v=full"
+        ).then(res => res.json()))
+    )
+  );
+
+/**
+ * returns true if no value is null or undefined
```

Comment:
Why do we need to know? I think thats where `??` is great.

Reply:
Would you say, it is okay to use nullish coalescing this way?
```ts
if(value ?? false) {
// ...
}
```

Reply:
Yes, I'd say so